### PR TITLE
Fix unbounded recursion

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/3672 (stack overflow on `where T : IComparable<T>` in public API analyzer when nullability analysis is enabled).
Closes https://github.com/dotnet/roslyn-analyzers/issues/3793 (adding a test)

### CODE REVIEW

**Comment by jcouv** [comment]
@mavasani @sharwell for review. Thanks

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 1]
```diff
@@ -832,7 +832,16 @@ public override bool VisitNamedType(INamedTypeSymbol symbol)
...
                     {
-                        if (Instance.Visit(typeArgument))
+                        // type parameters will already have been checked on the type defining them, so we can just do a shallow check
+                        if (typeArgument.TypeKind == TypeKind.TypeParameter)
+                        {
+                            if (typeArgument.IsReferenceType &&
+                                typeArgument.NullableAnnotation() == NullableAnnotation.None)
+                            {
+                                return true;
+                            }
```
Comment (sharwell) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:842]: ❔ Why no `else return false`?

Comment (jcouv) [reply src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:842]: We need to check the following type arguments (there may be more than one). Added a test to cover that scenario.

**Comment by jcouv** [comment]
@sharwell Good to go?

**Pull request merged**